### PR TITLE
Improve performance of `format_data()`

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- ~Tests Added~
- ~Documentation Added~
- ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR is an alternative approach to #726 & #727 only focusing on performance improvements in `format_data()` by switching from `pd.melt()` to `stack()`.

## Performance improvements

Current implementation

```
      ITEM_PATH      ITEM_VARIANT  TOTAL_TIME  CPU_USAGE    MEM_USAGE
0  test_profile  test_init[data0]    0.673276   0.203255    -1.433594
1  test_profile  test_init[data1]    9.140325   0.931676  1389.832031
2  test_profile  test_init[data2]   29.698939   0.932599  3407.539062
```

New implementation

```
      ITEM_PATH      ITEM_VARIANT  TOTAL_TIME  CPU_USAGE    MEM_USAGE
0  test_profile  test_init[data0]    0.786510   0.167053    -0.167969
1  test_profile  test_init[data1]    4.425207   0.850163   631.160156
2  test_profile  test_init[data2]   12.178936   0.942462  1482.765625
```

### CODE REVIEW

File: pyam/utils.py [hunk 1]
```diff
@@ -317,18 +330,20 @@ def convert_r_columns(c):
...
-            value_name="value",
-        )
+        # replace missing units by an empty string for user-friendly filtering
+        df.loc[df.unit.isnull(), "unit"] = ""
+
+        _validate_complete_index(df[index + REQUIRED_COLS + extra_cols])
+
+        # cast to long format, set
+        df.set_index(index + REQUIRED_COLS + extra_cols, inplace=True)
+        df = df.stack(dropna=True)
```
Comment (coroa) [inline pyam/utils.py:340]: Good find.

**Review by coroa — APPROVED** [review]
Sure, looks good!

My qualms, which the other PR #727 is addressing:
1. `format_data` is a maintenance burden, it's too long and difficult to parse. #727 split it out into four different functions: `convert_r_columns`, `knead_data` (not the best of names), `intuit_column_groups` and `transform_to_series`.
2. `format_data` here relies on creating a copy first. #727 avoids the copy in core.
3. The fast-case avoids reset_index with subsequent set_index, for correctly formatted timeseries()-like or inner _data-like representations, which is used all over the place from within `pyam` itself, but also for all the times, when you need to do calculations bypassing pyam.

I'd gladly add two more PRs on top of this one:
1. Refactor format_data into 5 different functions (above 4 + `_validate_complete_index`) -> #730 
2. Re-add multi-index fast-case -> #731

**Comment by gidden** [comment]
Thanks @danielhuppmann - I'll let @coroa finish off the review. In general I am happy especially if we minimize the read-in times of large AR6-like files as much as possible.

**Comment by danielhuppmann** [comment]
Thanks @coroa - fully agree with your qualms, happy to take this further and implement more of your suggestions! (Or even happier to review/merge this PR and you implement your suggestions on top of my changes). I was just worried that the current path-dependency of the existing PRs made it quite hard to review - and might make `format_data()` even more complicated and hard to maintain going forward.

File: pyam/utils.py [hunk 2]
```diff
@@ -371,6 +368,19 @@ def convert_r_columns(c):
...
+    """Validate that there are no nan's in the (index) columns"""
+    null_cells = df.isnull()
+    null_rows = null_cells.T.any()
+    if null_rows.any():
+        null_cols = null_cells.any()
+        cols = ", ".join(null_cols[null_cols].index)
+        raise_data_error(
+            f"Empty cells in `data` (columns: '{cols}')", df.loc[null_rows]
+        )
+    del null_rows
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/utils.py]: Done

**Comment by coroa** [comment]
I'm fine with merging this PR. (That's why i clicked *Approve* :))

**Pull request merged**